<p>raises an issue on a <code>For statement.</code> with a visitor on which you can control the visit.</p>
<h2>Noncompliant Code Example</h2>
<pre>
TO DO 
</pre>
<h2>Compliant Solution</h2>
<pre>
TO DO 
</pre>
